### ISSUE TITLE
Inconsistent results in `summary.survfit()` depending on the R version and the OS

### ISSUE THREAD

**Issue by DanChaltiel** [therneau/survival#273]
Hi,

First, thank you for this marvelous package.

I've experienced some inconsistency in `summary.survfit()` in my GitHub Actions tests.
This is part of a description package where I want survival variables to be described at all interesting times, so this may be a corner case.

Here is the minimal reproducible example:

``` r
df = data.frame(
  time = c(71.1, 75.7, 78.7, 79, 95.1, 108, 120.1, 120.3, 121, 140.8, 145, 146.7, 160,
           160, 167.6, 167.6, 225, 258, 275.8, 275.8, 275.8, 301, 304, 318, 350, 351,
           360, 360, 400, 440, 460, 472),
  event = rep(rep(c(1, 0), 6), c(6L, 1L, 2L, 1L, 1L, 1L, 2L, 7L, 1L, 3L, 1L, 6L)),
  group = rep(c("B", "A", "B", "A", "B", "A", "B", "A", "B", "A", "B"),
              c(9L, 1L, 1L, 1L, 2L, 3L, 1L, 1L, 7L, 2L, 4L))
)
df$time = round(df$time)
df$surv = survival::Surv(df$time, df$event)
fit = survival::survfit(surv~group, data=df)
times = sort(fit$time)

x = summary(fit, times=times, extend=TRUE)
a = data.frame(strata=x$strata, time=x$time, n.risk=x$n.risk, n.event=x$n.event, surv=x$surv)


a[a$time==141,]
#>     strata time n.risk n.event      surv
#> 8  group=A  141      8       0 1.0000000
#> 34 group=B  141     15       1 0.6640625

df[df$time==141,]
#>    time event group surv
#> 10  141     0     A 141+

packageVersion("survival")
#> [1] '3.7.0'
R.version.string
#> [1] "R version 4.4.1 (2024-06-14 ucrt)"
```

<sup>Created on 2024-08-25 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

The problem is that `n.event` at time 141 is sometimes 0 and sometimes 1 depending on the platform.

Here is the link to the GH Actions workflow: https://github.com/DanChaltiel/crosstable/actions/runs/10538974317
In my local installation, on Win10 with either R 4.3.1 or 4.4.1 and survival v3.7, I get `n.event==1`, as you can see on the reprex. 
However, on GH Actions, I get `n.event==0` on R-release on Windows, MacOS, and Ubuntu, and on R-old release on Ubuntu at least, and `n.event==1` on Ubuntu R-devel.

You can see the `n.event==0` result of the snapshot test for yourself in the falling jobs [here](https://github.com/DanChaltiel/crosstable/actions/runs/10538974317/job/29202052459#step:6:121), [here](https://github.com/DanChaltiel/crosstable/actions/runs/10538974317/job/29202052506#step:6:126), [here](https://github.com/DanChaltiel/crosstable/actions/runs/10538974317/job/29202052506#step:6:126), and [here](https://github.com/DanChaltiel/crosstable/actions/runs/10538974317/job/29202052506#step:6:126). The succeeding job has the same `n.event==1` result as on my local machine.

Feel free to ask for more information if needed.

EDIT: for testing purpose, running this exact same reprex on https://www.jdoodle.com/execute-r-online on any R version gives `n.event==0` like failing GH Actions and unlike my machine.

**Comment by therneau** [comment]
Please read the R FAQ, number 7.31 is what is almost certainly catching you.
On my machine for instance
 > print(140.8, digits=20)
[1] 140.80000000000001137

The value 140.8 does not have an exact representation in binary.   As you move that list of "time" values about in your test script, one can easily lose the last binary digit of precision.    There is a vignette on tied times in the survival package which goes into detail about one paticular roundoff error that occurs fairly frequently.

**Comment by DanChaltiel** [comment]
Thank you very much for your answer, I was not aware that this could be an issue.
For anyone interested in this thread, we are talking about [this vignette](https://cran.r-project.org/web/packages/survival/vignettes/tiedtimes.pdf).

However, the dataset has no tie, so I'm not sure that's the problem.
Moreover, the problem persists when you round the time to 141: [GHA results here](https://github.com/DanChaltiel/crosstable/actions/runs/10549243672/job/29224056103#step:6:260), and reprex has been updated.

Looking at `df`, I'm not sure I find my local reprex logical: the event at time 141 is a censoring (`event==0`) happening in group A, so getting `n.event==1` is a bit disturbing. I tend to agree more with GHA results than with the ones of my machine.

**Comment by DanChaltiel** [comment]
OK, I think there is something wrong.

On my Win10 machine with R version 4.4.1.
* With `survival 3.6.4` I get `n.event==0`
* With `survival 3.7.0` I get `n.event==1`
* With `survival 3.7.1` I get `n.event==1` (installed from GH using `pak::pak("therneau/survival")`)

I could not find version 3.6.5 on the CRAN or in GH releases so I couldn't test it.

I'm guessing `survival 3.7.0+` is installed on my machine and on the GHActions jobs machines that succeed and that it has some regressions causing this issue.

**Comment by therneau** [comment]
Use the sessionInfo() command to find out what version of the package is being used, after loading survival of course.

The vignette was triggered by case where time intervals that were tied can end up not tied due to finite precision.   The original purpose of the 'times' argument in summary.survfit routine was for printing out a more compact listing of a survival curve.  For example
```
> fit <- survfit(Surv(time/30.5, status) ~ sex, data=lung)  # time in months
> summary(fit, times= c(6,12,18,24))
Call: survfit(formula = Surv(time/30.5, status) ~ sex, data = lung)

                sex=1 
 time n.risk n.event survival std.err lower 95% CI upper 95% CI
    6     86      52   0.6225  0.0413        0.547        0.709
   12     35      33   0.3361  0.0434        0.261        0.433
   18     17      14   0.1897  0.0385        0.127        0.282
   24      7      10   0.0781  0.0276        0.039        0.156

                sex=2 
 time n.risk n.event survival std.err lower 95% CI upper 95% CI
    6     70      15    0.831  0.0399       0.7559        0.913
   12     30      21    0.526  0.0597       0.4215        0.658
   18     15       8    0.368  0.0630       0.2628        0.515
   24      5       7    0.156  0.0590       0.0743        0.328
```
The number of events shown in the second line is the cumulative number over the interval (6, 12] months.   Notice the round and square brackets.  Infinitesimal changes in survival times can cause a death that is exactly at 12 to end up in the wrong bin.   This is what I suspect is happening to you.   (I don't expect it to  happen above because the lung data set times are in days, and 30.5 has an exact representation in decimal.).

**Comment by therneau** [comment]
I have not been able to reproduce your results.  I don't have R on window available, but have used Ubuntu 22.04.4 LTS with R-devel on either  X-86-64 or a virtual machine on a M2 Mac, and  running  Rocky Line 8.10 on the department server.  Also R 4.3.2 with survival 3.6-1.    The GHActions reference is greek to me.

**Comment by DanChaltiel** [comment]
OK, I think I found the problem.

I was not careful enough when creating this `times` vector, which ends up having a duplicated final value.
This very unlikely event seems to be handled differently depending on `survival` version:
  
### survival v3.6.4
  
``` r
library(survival)
fit <- survfit(Surv(time/30.5, status) ~ 1, data=lung)
summary(fit, times= c(6,12,18,24,24))
# > Call: survfit(formula = Surv(time/30.5, status) ~ 1, data = lung)
# >
# >  time n.risk n.event survival std.err lower 95% CI upper 95% CI
# >     6    156      67    0.704  0.0304       0.6464        0.766
# >    12     65      54    0.409  0.0358       0.3447        0.486
# >    18     32      22    0.255  0.0344       0.1962        0.333
# >    24     12      17    0.107  0.0275       0.0645        0.177
# >    24     12       0    0.107  0.0275       0.0645        0.177
```

### survival v3.7.0 (wrong results)

``` r
library(survival)
fit <- survfit(Surv(time/30.5, status) ~ 1, data=lung)
summary(fit, times= c(6,12,18,24,24))
#> Call: survfit(formula = Surv(time/30.5, status) ~ 1, data = lung)
#> 
#>  time n.risk n.event survival std.err lower 95% CI upper 95% CI
#>     6    156       1    0.704  0.0304       0.6464        0.766
#>    12     65       1    0.409  0.0358       0.3447        0.486
#>    18     32       0    0.255  0.0344       0.1962        0.333
#>    24     12       1    0.107  0.0275       0.0645        0.177
#>    24     12       1    0.107  0.0275       0.0645        0.177
```

This seems to pertain to the introduction of the `dosum` argument, which changes the default behavior in certain conditions.

On my side, you can consider this issue resolved, as I will now call `unique()` on `times`.

However, note that there is an error in [summary.survfit.R#L110](https://github.com/therneau/survival/blob/master/R/summary.survfit.R#L110), where you should have written `is.logical(dosum)` instead of `logical(dosum)`. This error prevented me from using the `dosum` argument at all. 
Also, the last `else if` should be a normal `if` so that the check always happens. This would fix the current issue, as an error would have told me my `times` vector was misconstructed. 
With `dosum=TRUE` and bypassing the error, I find the old results back (but having an error is obviously better).

I'll make a PR right away.

**Comment by therneau** [comment]
Good catch.   I've fixed the error in my version, and added the following line of explanation to the help file:

...Alternately, the routine can return the number of events/censors/entry at
that time, \code{dosum=FALSE}.
This feature was added at the request of a user who essentially wanted to
use the \code{times} argument as a subscript to pick off selected rows
of the output, e.g., to select survival values corresponding to the
last follow-up times of a new set of observations.
The default for \code{dosum} is TRUE if the \code{times} vector is
strictly increasing and FALSE otherwise.
